### ISSUE TITLE
Tasks should be broken down into smaller pieces

### ISSUE THREAD

**Issue by mivade** [pennmem/ram_utils#15]
Too many things are happening in some pipeline tasks now. This means that if a long running step succeeds, there might be an overall failure from something trivial done afterwards requiring a full rerun of the task.

**Comment by mivade** [comment]
Ideally, tasks should be broken down into simple functions that produce a single output from a set of inputs. This gives us a number of advantages:

1. Inputs to a task are clearly defined via the arguments the function takes
2. Caching results is greatly simplified by using, e.g., `joblib.Memory` or a similar pattern
3. Testing of tasks is much easier: given a set of inputs, we know what the single output should look like
4. Parallelization is easier to achieve: independent chunks of work can be performed simultaneously and transparently via a scheduler like that in `dask.distributed` or the `joblib.Parallel` context

A simple example that can take advantage of each of these points is the existing `FREventPrepration` task. At present, it aggregates events from FR and catFR sessions, does some filtering, and finally concatenates the results. In a framework as described above, we would instead do something like the following:

```python
from joblib import Memory
from dask import delayed

mem = Memory(cachedir='/scratch/ramutils')

@mem.cache
def read_fr_events(subject):
    # read events...
    return events

@mem.cache
def read_catfr_events(subject):
    # read events...
    return events

@mem.cache
def combine_events(fr_events, cat_events):
    # concatenate...
    return events

fr_events = delayed(read_fr_events)("R0123E")
catfr_events = delayed(read_catfr_events)("R0123E")
all_events = delayed(combine_events)(fr_events, catfr_events)

# Dask's task scheduler automatically determines that fr_events and catfr_events can be
# run in parallel when it builds the task graph. Configuring how to run things in parallel, i.e.,
# using threads or processes or the cluster, is left as an exercise to the reader.
result = all_events.compute()
```

@zduey, @leondavis1, @maciekswat: thoughts?

**Comment by zduey** [comment]
Definitely in favor of this approach. Provided the tasks stay small, it will make our lives easier going forward. As an example, if we want to start caching report results in a database table, we simply add a new task, save_report_data(). Alternatively, if the source of some data changes, then it is an update to one function read_x(). Both are a huge improvement over what would currently need to be done.

What I'd want to know more about is how we can still have some pipline-y things such as continuing a pipeline from were it failed, rather than re-running everything. That is definitely a big "requirement" from my perspective in an updated version of our data processing pipeline. I imagine dask has some way of doing this, I just haven't done the research into it yet.

**Comment by mivade** [comment]
> such as continuing a pipeline from were it failed, rather than re-running everything

This is the point of caching. The decorator shown above causes the function to immediately return the cached value if it is present, otherwise it runs it and caches the result if successful. You can programmatically invalidate a function's cache with `function.clear()` or invalidate the entire cache with `mem.clear()`.

**Comment by zduey** [comment]
Ok, great
